File: soul-llm/soul_ai/rag.py
"""
Soul AI RAG Knowledge Engine for SOUL-LLM
==========================================
Lightweight BM25 / token matching knowledge retrieval for offline documents.
"""

import os
import re
import json
import math
from typing import List, Dict, Any

KNOWLEDGE_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "data", "knowledge"))
os.makedirs(KNOWLEDGE_DIR, exist_ok=True)


class SimpleRAG:
    def __init__(self, storage_dir: str = KNOWLEDGE_DIR):
        self.storage_dir = storage_dir
        os.makedirs(self.storage_dir, exist_ok=True)
        self.documents: List[Dict[str, Any]] = []
        self._load_documents()

    def _tokenize(self, text: str) -> List[str]:
        return [w.lower() for w in re.findall(r"\w+", text)]

    def _load_documents(self):
        self.documents = []
        for fname in os.listdir(self.storage_dir):
            if fname.endswith(".json"):
                fpath = os.path.join(self.storage_dir, fname)
                try:
                    with open(fpath, "r", encoding="utf-8") as f:
                        data = json.load(f)
                        self.documents.append(data)
                except Exception:
                    pass

    def add_document(self, title: str, content: str) -> Dict[str, Any]:
        """Indexes and saves a document into the offline knowledge base."""
        doc_id = re.sub(r"[^\w\-]", "_", title.lower())
        doc_data = {
            "id": doc_id,
            "title": title,
            "content": content,
            "tokens": self._tokenize(content),
            "size": len(content)
        }
        fpath = os.path.join(self.storage_dir, f"{doc_id}.json")
        with open(fpath, "w", encoding="utf-8") as f:
            json.dump(doc_data, f, indent=2)

        # Update in-memory
        self.documents = [d for d in self.documents if d["id"] != doc_id]
        self.documents.append(doc_data)
        return {"status": "success", "id": doc_id, "title": title, "tokensCount": len(doc_data["tokens"])}

    def search(self, query: str, top_k: int = 3) -> List[Dict[str, Any]]:
        """Search documents using term frequency ranking."""
        query_tokens = self._tokenize(query)
        if not query_tokens or not self.documents:
            return []

        scored = []
        for doc in self.documents:
            doc_tokens = doc.get("tokens", [])
            if not doc_tokens:
                continue
            doc_len = len(doc_tokens)
            score = 0.0
            for qt in query_tokens:
                tf = doc_tokens.count(qt)
                if tf > 0:
                    score += (tf / doc_len) * (1.0 + math.log(1 + 10 / max(1, len(self.documents))))
            if score > 0:
                scored.append({
                    "title": doc["title"],
                    "score": round(score, 4),
                    "snippet": doc["content"][:300] + ("..." if len(doc["content"]) > 300 else "")
                })

        scored.sort(key=lambda x: x["score"], reverse=True)
        return scored[:top_k]

    def list_documents(self) -> List[Dict[str, Any]]:
        return [
            {"id": d["id"], "title": d["title"], "size": d["size"], "tokens": len(d.get("tokens", []))}
            for d in self.documents
        ]


rag_engine = SimpleRAG()

File: agent/rag.js
const fs = require('fs');
const path = require('path');

const IS_SERVERLESS = !!process.env.VERCEL;
const KNOWLEDGE_DIR = path.resolve(__dirname, '..', 'storage', 'knowledge');
const INDEX_FILE = path.resolve(__dirname, '..', 'storage', 'rag_index.json');

if (!IS_SERVERLESS && !fs.existsSync(KNOWLEDGE_DIR)) {
  fs.mkdirSync(KNOWLEDGE_DIR, { recursive: true });
}

// Simple fast tokenizer for offline ranking
function tokenize(text) {
  return text
    .toLowerCase()
    .replace(/[^\w\s]/g, ' ')
    .split(/\s+/)
    .filter(t => t.length > 1);
}

// Split document text into chunks with overlap
function chunkText(text, chunkSize = 400, overlap = 80) {
  const words = text.split(/\s+/);
  const chunks = [];
  let i = 0;
  while (i < words.length) {
    const chunkWords = words.slice(i, i + chunkSize);
    chunks.push(chunkWords.join(' '));
    i += chunkSize - overlap;
  }
  return chunks.length ? chunks : [text];
}

class OfflineRagEngine {
  constructor() {
    this.documents = [];
    this.chunks = [];
    this.loadIndex();
  }

  loadIndex() {
    try {
      if (fs.existsSync(INDEX_FILE)) {
        const data = JSON.parse(fs.readFileSync(INDEX_FILE, 'utf8'));
        this.documents = data.documents || [];
        this.chunks = data.chunks || [];
      } else {
        this.rebuildFromFolder();
      }
    } catch {
      this.documents = [];
      this.chunks = [];
    }
  }

  saveIndex() {
    try {
      fs.writeFileSync(INDEX_FILE, JSON.stringify({
        documents: this.documents,
        chunks: this.chunks
      }, null, 2));
    } catch (err) {
      console.error('Failed to save RAG index:', err);
    }
  }

  rebuildFromFolder() {
    this.documents = [];
    this.chunks = [];
    if (!fs.existsSync(KNOWLEDGE_DIR)) return;

    const files = fs.readdirSync(KNOWLEDGE_DIR);
    for (const file of files) {
      const fullPath = path.join(KNOWLEDGE_DIR, file);
      const stat = fs.statSync(fullPath);
      if (stat.isFile()) {
        const content = fs.readFileSync(fullPath, 'utf8');
        this.addDocument(file, content, false);
      }
    }
    this.saveIndex();
  }

  addDocument(filename, content, save = true) {
    // Remove if already exists
    this.documents = this.documents.filter(d => d.filename !== filename);
    this.chunks = this.chunks.filter(c => c.source !== filename);

    const docMeta = {
      filename,
      charCount: content.length,
      indexedAt: new Date().toISOString()
    };
    this.documents.push(docMeta);

    // Chunk content
    const textChunks = chunkText(content);
    for (let idx = 0; idx < textChunks.length; idx++) {
      const chunkTextStr = textChunks[idx];
      const tokens = tokenize(chunkTextStr);
      const tf = {};
      for (const token of tokens) {
        tf[token] = (tf[token] || 0) + 1;
      }
      this.chunks.push({
        id: `${filename}_chunk_${idx}`,
        source: filename,
        chunkIndex: idx,
        text: chunkTextStr,
        tokensCount: tokens.length,
        tf
      });
    }

    if (save) {
      this.saveIndex();
    }
  }

  removeDocument(filename) {
    this.documents = this.documents.filter(d => d.filename !== filename);
    this.chunks = this.chunks.filter(c => c.source !== filename);
    const fullPath = path.join(KNOWLEDGE_DIR, filename);
    if (fs.existsSync(fullPath)) {
      fs.unlinkSync(fullPath);
    }
    this.saveIndex();
    return { success: true, removed: filename };
  }

  search(query, topK = 3) {
    if (!this.chunks.length) {
      return {
        query,
        count: 0,
        results: [],
        message: 'No documents in knowledge base. Add files in storage/knowledge or via dashboard.'
      };
    }

    const queryTokens = tokenize(query);
    if (!queryTokens.length) {
      return { query, count: 0, results: [] };
    }

    // BM25-style scoring
    const N = this.chunks.length;
    const avgDocLen = this.chunks.reduce((acc, c) => acc + c.tokensCount, 0) / N;
    const k1 = 1.5;
    const b = 0.75;

    const scores = this.chunks.map(chunk => {
      let score = 0;
      const docLen = chunk.tokensCount;

      for (const token of queryTokens) {
        // Document frequency
        const df = this.chunks.filter(c => c.tf[token]).length;
        if (df > 0) {
          const idf = Math.log((N - df + 0.5) / (df + 0.5) + 1);
          const f = chunk.tf[token] || 0;
          const termScore = idf * ((f * (k1 + 1)) / (f + k1 * (1 - b + b * (docLen / avgDocLen))));
          score += termScore;
        }
      }

      return {
        score,
        source: chunk.source,
        chunkIndex: chunk.chunkIndex,
        text: chunk.text
      };
    });

    const ranked = scores
      .filter(s => s.score > 0)
      .sort((a, b) => b.score - a.score)
      .slice(0, topK);

    return {
      query,
      count: ranked.length,
      results: ranked
    };
  }

  getStats() {
    return {
      totalDocuments: this.documents.length,
      totalChunks: this.chunks.length,
      documents: this.documents
    };
  }
}

const ragEngine = new OfflineRagEngine();

module.exports = {
  ragEngine,
  KNOWLEDGE_DIR
};
